onboarding/mlsys/1-backprop.md

```diff
@@ -0,0 +1,414 @@
+
+- `(x=1, ylabel=1)`
+- `(x=2, ylabel=2)`
+- `(x=3, ylabel=3)`
+- `(x=4, ylabel=4)`
+
+## The Math Problem
+
+What is the best value of `m` to make our model (`ypred = m*x`) most-accurately
+predict our dataset?
+
+### Q1:
+
+The answer is `m=1`.  Looking at the dataset, explain why `m=1` is correct
+before proceeding with further math.
+
+## What does "most-accurately" mean, really?
+
+For this model and dataset, there is an exact answer (`m`=1), but usually there
+isn't.  So usually instead of seeking an exact answer, we are really seeking to
+minimize a function, called the "loss", that mathematically describes the
+goodness-of-fit between our model and the data.  There are lots of loss
+functions, but we're going to use the least-squares error function, sometimes
+abbreviated LSE.  It is defined as the sum of the squares of the error between
+our labels (the `y` values of our dataset) and our predictions (the output `y`
+of our `y=m*x` model).
+
+So for data points `(x0, ylabel0)`, `(x1, ylabel1)` and corresponding
+predictions `ypred0`, `ypred1`, you would calculate the LSE as:
+
+    loss = (ylabel0 - ypred0)**2 + (ylabel1 - ypred1)**2
+
+### Q2:
+
+Complete the following table, calculate the loss for each point of our dataset
+if we set `m=2` in our model:
+
+|  x  | ylabel | ypred | `(ylabel - ypred)**2` |
+| --- | ------ | ----- | --------------------- |
+|  1  |   1    |       |                       |
+|  2  |   2    |       |                       |
+|  3  |   3    |       |                       |
+|  4  |   4    |       |                       |
+
+Write a function for calculating the LSE loss for a single `(x, ylabel)` data
+pair.  It should take `m` and a single `(x, ylabel)` pair as inputs and return
+the loss.
+
+Then write a function for calculating the LSE loss for a whole dataset.  It
+should take `m` and a list of `(x, ylabel)` pairs and return the LSE loss.
+
+Is it possible to write the second function by making multiple calls to the
+first function and combining the results?  Why or why not?
+
+## Minimizing the loss
+
+We have the loss, but what do we do to minimize it?  The answer is called
+"Gradient Descent".  We basically follow the curve of our loss until we get to
+the minimum of the loss graph:
+
+![gradient descent](gradient_descent.jpg "Gradient Descent")
+
+That image is pulled from from [this youtube video series](
+https://www.youtube.com/watch?v=b4Vyma9wPHo) which explains a lot of the
+details here in greater detail, if you want that.
+
+The basic idea is that we look at the slope of the loss curve (or the "gradient"
+of the loss curve if our loss is more than two dimensional), and try to take a
+step in the direction of the minimum.
+
+The direction of our step is based on the sign of the slope.  If the slope is
+positive, we know the minimum must be to our left, so we move in the negative
+direction (decreasing `m`).
+
+The size of our step is based on the magnitude of the slope.  If the slope is
```

Comment:
> The size of our step is based on the magnitude of the slope.
I would probably introduce the `lr` here, since the size of the step is based on both the gradient magnitude and `lr`.   "The size of our step is based both on the magnitude of the slope and a tunable parameter known as the learning rate." or similar